assets/javascripts/discourse/components/d-progress-bar.js

```diff
@@ -0,0 +1,99 @@
+import Component from "@ember/component";
+import { cancel, later, schedule } from "@ember/runloop";
+
+const STORE_LOADING_TIMES = 5;
+const DEFAULT_LOADING_TIME = 0.3;
+const MIN_LOADING_TIME = 0.1;
+const STILL_LOADING_DURATION = 2;
+
+export default Component.extend({
+  tagName: "",
+  isLoading: false,
+  key: null,
+
+  init() {
+    this._super(...arguments);
+
+    this.loadingTimes = [DEFAULT_LOADING_TIME];
+    this.set("averageTime", DEFAULT_LOADING_TIME);
+    this.i = 0;
+    this.scheduled = [];
+  },
+
+  cancelScheduled() {
+    this.scheduled.forEach((s) => cancel(s));
+    this.scheduled = [];
+  },
+
+  didReceiveAttrs() {
+    this._super(...arguments);
+
+    if (!this.key) {
+      return;
+    }
+
+    if (this.isLoading) {
+      this.start();
+    } else {
+      this.end();
+    }
+  },
+
+  get container() {
+    return document.getElementById(this.key);
```

Comment:
I am not seeing where `key` is set. It's probably me, but it seems odd.

Reply:
It's when you defined the component, to have a uniq key if you have multiple bars

Reply:
{{d-progress-bar key="cats-loader" isLoading=(not isDone)}}